core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -0,0 +1,436 @@
+                ")";
+        }
+    }
+
+    /**
+     * The InFlightState is used to track the state and delivery count of a record that has been
+     * fetched from the leader. The state of the record is used to determine if the record should
+     * be re-deliver or if it can be acknowledged or archived.
+     */
+    private static class InFlightState {
+        /**
+         * The state of the fetch batch records.
+         */
+        private RecordState state;
+        /**
+         * The number of times the records has been delivered to the client.
+         */
+        private int deliveryCount;
+        /**
+         * The member id of the client that is fetching/acknowledging the record.
+         */
+        private String memberId;
+
+        InFlightState(RecordState state, int deliveryCount, String memberId) {
+            this.state = state;
+            this.deliveryCount = deliveryCount;
+            this.memberId = memberId;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(state, deliveryCount, memberId);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            InFlightState that = (InFlightState) o;
+            return state == that.state && deliveryCount == that.deliveryCount && memberId.equals(that.memberId);
+        }
+
+        @Override
+        public String toString() {
+            return "InFlightState(" +
+                " state=" + state.toString() +
+                ", deliveryCount=" + deliveryCount +
+                ", memberId=" + memberId +
+                ")";
+        }
+    }
+
+    /**
+     * The AcknowledgementBatch containing the fields required to acknowledge the fetched records.
+     */
+    public static class AcknowledgementBatch {
+
+        private final long firstOffset;
+        private final long lastOffset;
+        private final List<Byte> acknowledgeTypes;
+
+        public AcknowledgementBatch(long firstOffset, long lastOffset, List<Byte> acknowledgeTypes) {
+            this.firstOffset = firstOffset;
+            this.lastOffset = lastOffset;
+            this.acknowledgeTypes = acknowledgeTypes;
+        }
+
+        public long firstOffset() {
+            return firstOffset;
+        }
+
+        public long lastOffset() {
+            return lastOffset;
+        }
+
+        public List<Byte> acknowledgeTypes() {
+            return acknowledgeTypes;
+        }
+
+        @Override
+        public String toString() {
+            return "AcknowledgementBatch(" +
+                " firstOffset=" + firstOffset +
+                ", lastOffset=" + lastOffset +
+                ", acknowledgeTypes=" + ((acknowledgeTypes == null) ? "" : acknowledgeTypes) +
+                ")";
+        }
+    }
+}
```

Comment:
nit: We usually have an empty line at the end.

Reply:
Done, thanks.